### ISSUE TITLE
Gradient-aware spline upscaling as a zoom-in alternative

### ISSUE THREAD

**Issue by nk4517** [1207koo/gaussian_blending#3]
### Title: Gradient-aware spline upscaling as a zoom-in alternative

In Section C.4 you note that zoom-in remains an open problem — high-frequency noise on specular objects and limited detail recovery from low-res training data. Posting this as potentially relevant information.

**Paper:** Niedermayr, Neuhauser, Westermann (TU Munich) — "Lightweight Gradient-Aware Upscaling of 3D Gaussian Splatting Images", [arXiv:2503.14171v2](https://arxiv.org/abs/2503.14171)

**Unofficial PyTorch/CUDA implementation (upscaler only, differentiable):**
https://github.com/nk4517/lwga_upscaler

The idea: compute analytical image-space gradients (∂I/∂x, ∂I/∂y, ∂²I/∂x∂y) from Gaussian footprints during the forward pass, then use them for bicubic spline interpolation instead of finite differences. Produces sharper upscaling at ~1.8ms overhead for 4K at 3x — no learned priors, no hallucination, temporally stable.

These two methods address different zoom regimes rather than being combined in a single pipeline:

- **Scale < 1 (zoom-out):** Gaussian Blending handles dilation/erosion via spatial transmittance tracking.
- **Scale 1x–10x (zoom-in):** render at 1x resolution with analytical gradients, then upscale ROI to target resolution. No blending overhead, and the upscaler itself is O(1) in memory — avoids the massive GPU cost of rendering at 10x.

The attached video demonstrates this — a continuous zoom sweep where the two regimes switch at 1x. The video was generated using a 2D Gaussian image model, but the approach works without modification for any splatting representation that exposes covariance/conics (3DGS, GaussianImage, etc.), and with some additional work — for any renderer where analytical screen-space gradients can be derived.

A working demo has been merged into the [LIG (Large Images are Gaussians)](https://github.com/HKU-MedAI/LIG) repository — see [PR #3](https://github.com/HKU-MedAI/LIG/pull/3), which adds gradient-aware upscaling with differentiable training for 2D Gaussian image representation. There is also working (unpublished) code for the zoom-out direction based on Analytic-Splatting (in the 2D Gaussian image case, blending uses a weighted sum rather than accumulated transmittance, so there is no need for transmittance window management as in Gaussian Blending).

https://github.com/user-attachments/assets/2ba53f2a-f727-4996-82c2-6fbda7cfb099

<img width="512" height="341" alt="Image" src="https://github.com/user-attachments/assets/8beaf14c-072a-44b8-a1ea-2fcd138c4ee6" />

<img width="512" height="341" alt="Image" src="https://github.com/user-attachments/assets/ebd89e73-f91f-42a9-8ee2-08fc399860ef" />
<img width="512" height="341" alt="Image" src="https://github.com/user-attachments/assets/7c5c394b-528a-48e6-a3a0-a1b33fd41b43" />

**Comment by nk4517** [comment]
One more note on the upscaler's behavior with blur:

LWGA is completely agnostic to source image blur — if the Gaussian model has learned a blurry region (e.g., from defocus, motion blur, or low training resolution), the upscaler will reconstruct that blur with full sub-pixel fidelity, to the extent the Gaussian budget allows. There is no sharpening bias or detail hallucination — the spline simply fits whatever the Gaussians actually represent, blurry or not.

A related property emerges when training through the upscaler at reduced resolution (e.g., 1/2). The Gaussians must simultaneously produce a pixel-perfect image at the native render resolution *and* a pixel-perfect match to the ground truth after upscaling. These are two different sampling grids with different pixel boundaries, so the Gaussians cannot simply memorize pixel values — they are forced to approximate the underlying continuous signal geometry instead. This effectively removes source pixelation artifacts and produces smooth, visually unpixelated zooms at 4x–8x (though without synthesizing detail absent from the original, of course).
